File: problems/leetcode/0200/210/210.py
# 210. Course Schedule II
# https://leetcode.com/problems/course-schedule-ii/description/

from collections import defaultdict, deque
from typing import *

class Solution:
    def findOrder(self, numCourses: int, prerequisites: List[List[int]]) -> List[int]:
        indegrees = [0 for _ in range(numCourses)]
        outdegrees = defaultdict(set[int])

        for src, dst in prerequisites:
            indegrees[src] += 1
            outdegrees[dst].add(src)

        q = deque()
        for i in range(numCourses):
            if indegrees[i] == 0:
                q.append(i)

        courseOrder = []
        while q:
            currentCourses = list(q)
            while q:
                q.popleft()
            courseOrder += currentCourses

            for course in currentCourses:
                for nextCourse in outdegrees[course]:
                    indegrees[nextCourse] -= 1
                    if indegrees[nextCourse] == 0:
                        q.append(nextCourse)

        return courseOrder if len(courseOrder) == numCourses else []

File: problems/leetcode/0200/210/210.cpp
// 210. Course Schedule II
// https://leetcode.com/problems/course-schedule-ii/

using namespace std;

#include <map>
#include <set>
#include <vector>

class Solution {
public:
  vector<int> findOrder(int numCourses, vector<vector<int>> &prerequisites) {
    vector<int> answer;
    map<int, set<int>> innerVertices, outerVertices;
    set<int> initialCandidates;

    for (int i = 0; i < numCourses; i++) {
      initialCandidates.insert(i);
    }

    for (auto &prerequisite : prerequisites) {
      int earlier = prerequisite[1], later = prerequisite[0];
      innerVertices[later].insert(earlier);
      outerVertices[earlier].insert(later);
      initialCandidates.erase(later);
    }

    for (auto &course : initialCandidates) {
      answer.push_back(course);
    }

    int i = 0;
    while (i < answer.size()) {
      int course = answer[i];

      for (auto &nextCourse : outerVertices[course]) {
        innerVertices[nextCourse].erase(course);
        if (innerVertices[nextCourse].size() == 0) {
          answer.push_back(nextCourse);
        }
      }

      i++;
    }

    if (answer.size() != numCourses) {
      answer.clear();
    }

    return answer;
  }
};
